pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -369,27 +381,36 @@ def extract_object_metadata_from_headers(headers):
     return meta_headers
 
 
-def extract_container_metadata_from_headers(headers):
+def extract_container_metadata_from_headers(req):
     """
-    Find and return the key/value pairs containing object metadata.
+    Find and return the key/value pairs containing container metadata.
 
-    This tries to do the same thing as the Swift object server: save only
+    This tries to do the same thing as the Swift container server: save only
     relevant headers. If the user sends in "X-Fungus-Amungus: shroomy" in
     the PUT request's headers, we'll ignore it, just like plain old Swift
     would.
 
-    :param headers: request headers (a dictionary)
+    :param req: a swob Request
 
-    :returns: dictionary containing object-metadata headers
+    :returns: dictionary containing container-metadata headers
     """
     meta_headers = {}
-    for header, value in headers.items():
+    for header, value in req.headers.items():
         header = header.title()
 
-        if (header.startswith("X-Container-Meta-") or
+        if ((header.startswith("X-Container-Meta-") or
                 header.startswith("X-Container-Sysmeta-") or
-                header in SPECIAL_CONTAINER_METADATA_HEADERS):
+                header in SPECIAL_CONTAINER_METADATA_HEADERS) and
+                (req.environ.get('swift_owner', False) or
+                 header not in SWIFT_OWNER_HEADERS)):
             meta_headers[header] = value
+        if header.startswith("X-Remove-"):
+            header = header.replace("-Remove", "", 1)
+            if ((header.startswith("X-Container-Meta-") or
+                    header in SPECIAL_CONTAINER_METADATA_HEADERS) and
+                    (req.environ.get('swift_owner', False) or
+                     header not in SWIFT_OWNER_HEADERS)):
```

Comment:
Shouldn't this include sysmeta as well?

Hmm... or does the noauth-proxy's gatekeeper drop it regardless of what we do here?

Reply:
the x-remove-* handling in swift is (as I understand it) for clients that can't get empty header values to the proxy. sysmeta can and should always be deleted by simply setting an empty value. So the behaviour above should  mimic exactly what swift proxy-server does. (Note that non-remove sysmeta is passed).

AFAICT these metadata don't get passed as headers to the noauth proxy - if they did then there would be issues with swift_owner being False (since there is no auth to set swift_owner). In fact my first reaction to some of these bug was "how come the noauth proxy isn't filtering/transforming headers?". However I don't yet have a good understanding of what happens on the proxyfs side.